runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkBundleFactory.java

```diff
@@ -0,0 +1,42 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.runners.flink;
+
+import org.apache.beam.runners.fnexecution.artifact.ArtifactSource;
+import org.apache.beam.runners.fnexecution.control.JobBundleFactory;
+import org.apache.beam.runners.fnexecution.provisioning.JobInfo;
+
+/**
+ * A simple {@link JobBundleFactory} for Flink operators. Uses Docker for underlying environment
+ * management.
+ */
+public class FlinkBundleFactory {
+
+  public static FlinkBundleFactory getInstance() {
+    return Holder.INSTANCE;
+  }
+
+  /** Gets or creates a {@link JobBundleFactory}. */
+  public JobBundleFactory getJobBundleFactory(JobInfo jobInfo, ArtifactSource artifactSource) {
+    throw new UnsupportedOperationException();
+  }
+
+  private static class Holder {
```

Comment:
Is this class necessary?

Reply:
Removed.

Note: I originally thought this class might do more work and wanted lazy initialization. However, given that the bundle factory itself must have TaskManager-local state, this cannot be passed directly to the `FlinkExecutableStage` at construction. Instead, any heavy lifting will happen in the `SerializableSupplier` that produces this. `getInstance` will probably change to return a `SerializableSupplier<FlinkBundleFactory>`. This is effectively a FactoryFactory(FactoryFactoryFactory).

It's unfortunate we need this many layers of indirection for object creation, but it's necessary for unit testing and distributed execution.